cohortextractor/query_engines/mssql_lib.py

```diff
@@ -0,0 +1,339 @@
+import contextlib
+import hashlib
+import logging
+import os
+import secrets
+import time
+
+import sqlalchemy
+
+
+log = logging.getLogger(__name__)
+
+
+def write_query_to_table(table, query):
+    """
+    Write the results of `query` into `table` (both SQLAlchemy objects)
+    """
+    # This is a bit of a hack but works OK. We want to be able to take
+    # an arbitrary select query and generate the SQL:
+    #
+    #   SELECT * INTO some_temporary_table FROM (some_select_query) AS some_alias
+    #
+    # which is the MSSQL-specific syntax for writing the results of a
+    # query directly into a temporary table. We can trick SQLAlchemy
+    # into generating this for us by giving it a literal column named
+    # "* INTO table_name".
+    into_table = sqlalchemy.literal_column(f"* INTO {table.name}")
+    return sqlalchemy.select(into_table).select_from(query.alias())
+
+
+@contextlib.contextmanager
+def fetch_results_in_batches(
+    engine,
+    queries,
+    temp_table_prefix=None,
+    key_column="patient_id",
+    **batch_fetch_config,
+):
+    """
+    Context manager which takes a list of queries and executes them, returning
+    an iterator over the results of the final query.
+
+    Behind the scenes this writes the results to a temporary table and then
+    downloads the results in batches, retrying after any errors. This allows us
+    to workaround flakiness in our connection to the database.
+
+    Args:
+        engine: sqlalchemy.Engine instance
+
+        queries: list of sqlalchemy queries, with the final being a Select
+
+        temp_table_prefix: prefix to use when generating temporary table names
+            use a leading `#` character for session-scoped temporary tables,
+            use the `DatabaseName..table_prefx` syntax to store tables in
```

Comment:
Is the `..` a typo?

Reply:
Looks like it's not.  Might be worth a brief comment for the uninitiated?

Reply:
something, something mssql something ...

I'll explicitly refer to it as the double-dot syntax in the comment so it's clear it's not a typo.